bundles/org.openhab.binding.sony/src/main/java/org/openhab/binding/sony/internal/net/SocketChannelSession.java

```diff
@@ -0,0 +1,308 @@
+    /** The logger */
+    private final Logger logger = LoggerFactory.getLogger(SocketChannelSession.class);
+
+    /** The host/ip address to connect to */
+    private final String host;
+
+    /** The port to connect to */
+    private final int port;
+
+    /**
+     * The actual socket being used. Will be null if not connected
+     */
+    private final AtomicReference<@Nullable SocketChannel> socketChannel = new AtomicReference<>();
+
+    /** The responses read from the {@link #responseThread}. */
+    private final BlockingQueue<Object> responses = new ArrayBlockingQueue<Object>(50);
+
+    /** The {@link SocketSessionListener} that the {@link #dispatchingThread} will call. */
+    private final List<SocketSessionListener> sessionListeners = new CopyOnWriteArrayList<>();
+
+    /** The thread dispatching responses - will be null if not connected. */
+    private @Nullable Thread dispatchingThread = null;
+
+    /** The thread processing responses - will be null if not connected. */
+    private @Nullable Thread responseThread = null;
+
+    /**
+     * Creates the socket session from the given host and port.
+     *
+     * @param host a non-null, non-empty host/ip address
+     * @param port the port number between 1 and 65535
+     */
+    public SocketChannelSession(final String host, final int port) {
+        Validate.notEmpty(host, "host cannot be null");
+
+        if (port < 1 || port > 65535) {
+            throw new IllegalArgumentException("Port must be between 1 and 65535");
+        }
+        this.host = host;
+        this.port = port;
+    }
+
+    @Override
+    public void addListener(final SocketSessionListener listener) {
+        Objects.requireNonNull(listener, "listener cannot be null");
+        sessionListeners.add(listener);
+    }
+
+    @Override
+    public void clearListeners() {
+        sessionListeners.clear();
+    }
+
+    @Override
+    public boolean removeListener(final SocketSessionListener listener) {
+        Objects.requireNonNull(listener, "listener cannot be null");
+        return sessionListeners.remove(listener);
+    }
+
+    @Override
+    public void connect() throws IOException {
+        connect(2000);
+    }
+
+    @Override
+    public void connect(final int timeout) throws IOException {
+        disconnect();
+
+        final SocketChannel channel = SocketChannel.open();
+        channel.configureBlocking(true);
+
+        logger.debug("Connecting to {}:{}", host, port);
+        channel.socket().connect(new InetSocketAddress(host, port), timeout);
+
+        socketChannel.set(channel);
+
+        responses.clear();
+
+        dispatchingThread = new Thread(new Dispatcher());
+        responseThread = new Thread(new ResponseReader());
+
+        dispatchingThread.setDaemon(true);
+        responseThread.setDaemon(true);
+
+        dispatchingThread.start();
+        responseThread.start();
+    }
+
+    @Override
+    public void disconnect() throws IOException {
```

Comment:
A lot of NPE would happen if this got called by two threads at the same time.

Reply:
Yuck - that was one of the very first classes I wrote and it certainly has issues with multithreading.  Should be fixed in the next commit